### PR TITLE
SEARCH-461 cache only on leaders and SEARCH-440

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

And fix truncate related issue: https://github.com/iresearch-toolkit/iresearch/pull/496

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1253
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchDataStore.cpp [hunk 1]
```diff
@@ -869,6 +879,27 @@ Result IResearchDataStore::commitUnsafeImpl(
 
       commitLock.lock();
     }
+
+#ifdef USE_ENTERPRISE
+    bool const reopenColumnstore = [&] {
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:884]: Why do we need lamda here? It is called only once. I suggest to just calculate boolean here.

**Review by Dronplane — CHANGES_REQUESTED** [review]

File: arangod/IResearch/IResearchDataStore.cpp (hunk 1 above)
Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:884]: To make it const + scope looks good for me

File: tests/js/client/chaos/test-chaos-load-common.inc [hunk 2]
```diff
@@ -416,6 +416,7 @@ function BaseChaosSuite(testOpts) {
             if (d >= 0.98 && testOpts.withTruncate) {
               log("RUNNING TRUNCATE");
               c.truncate();
+              // query("FOR doc IN " + c.name() + " REMOVE doc IN " + c.name());
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply tests/js/client/chaos/test-chaos-load-common.inc:419]: Not really, I noticed them, but thought they can be useful, so why not

[the remark this answers is not in the document]
Comment (Dronplane) [reply tests/js/client/chaos/test-chaos-load-common.inc:419]: Let's not put unrelated changes in the PR please.  More to say - this is completely temporary code to debug another issue. It should not be there  anyway.

[the remark this answers is not in the document]
Comment (MBkkt) [reply tests/js/client/chaos/test-chaos-load-common.inc]: Fixed

File: arangod/IResearch/IResearchDataStore.cpp [hunk 3]
```diff
@@ -877,47 +908,39 @@ Result IResearchDataStore::commitUnsafeImpl(
                            "ArangoSearch index '",
                            index().id().id(), "'")};
     }
-    auto const lastTickBeforeCommit = engineSnapshot->tick();
-    TRI_ASSERT(_lastCommittedTick <= lastTickBeforeCommit);
-    absl::Cleanup commitGuard = [&, lastCommittedTick =
-                                        _lastCommittedTick]() noexcept {
-      _lastCommittedTick = lastCommittedTick;
+    auto const beforeCommit = engineSnapshot->tick();
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:911]: IMO `lastTickBeforeCommit` was better

**Review by Dronplane — APPROVED** [review]
LGTM

**Review by joerg84 — APPROVED** [review]
It would have been nice to separate the fix from the other improvements for 3.11, but ok

**Pull request merged**